### ISSUE TITLE
Cache SemanticModel Instances

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#39840]
We currently don't cache SemanticModel instances today. This is, in part, because calls to methods on the model would only cause binding on an individual statement. However, with C# 8 and nullable, we now have to bind the entire body of a member in order to provide semantic information when nullable is enabled. Some existing analyzers have taken an implicit dependency on the previously-true fact that SemanticModels are cheap to create and ask a single question of, but are now seeing performance regressions because this is no longer true. We should consider lazily caching SyntaxTreeSemanticModels on the Compilation, when asked for, which will improve the performance of these analyzers, especially since the guidance around `GetSemanticModel` for a single question has now changed.

/cc @agocke

**Comment by agocke** [comment]
Note: not a regression, because this is only a problem on C# 8 code that uses nullable. But this could be a performance improvement for nullable code.

**Comment by CyrusNajmabadi** [comment]
We would need ot be *very* careful here.  Non-caching of this is a specific design decision from Roslyn from teh start.  In essence the semantic model *is* the cache.  The person who creates it is the one who should be responsible for determining how long they need this cache for.

 >Some existing analyzers have taken an implicit dependency on the previously-true fact that SemanticModels are cheap to create and ask a single question of

These analyzers should be fixed.  It's never been the case AFAICT, that you could presume a semantic model (or asking a semantic question on a model) was cheap.  For example, there were lots of cases where you might ask a question, and tons of binding might be necessary (i.e. in lambda scenarios and others).  So this was a spurious assumption in the first place.

**Comment by agocke** [comment]
@CyrusNajmabadi I agree in general. Caching semantic model instances is dangerous because it could cause you to hold on to bound nodes.

What I wonder is if we could try to share semantic models at a higher level from the symbol action analyzers, if it's the case that there isn't really a benefit to providing a semantic model with a finer-grained level than method body.

**Comment by CyrusNajmabadi** [comment]
> What I wonder is if we could try to share semantic models at a higher level from the symbol action analyzers,

That would be very interesting.  The reason right now that isn't done is because a symbol can be declared across *many* semantic models (i.e. partials).  I can imagine the analyzer being smart and exposin all the semantic models that are needed for a particular symbol *with* that symbol.

The actual order of symbol processing would then be written to hold onto semantic models for as long as necessary while processing, but no longer.

**Comment by 333fred** [comment]
@jaredpar I removed the bug label. The current behavior is by-design, changing it would be a feature request.

**Comment by CyrusNajmabadi** [comment]
Have discovered from @mavasani that the compiler doesn't use the same semantic model it already has when a user registers a SemanticModelAnalysis action.  It will create a fresh one, causing all binding in that action to recompute the results the compiler already has computed for the analysis it is doing.

This seems easily solvable.  If there are semantic model actions requested, the compiler should just make semantic models for the files, then analyze all semantics in those files using that same semantic model.

**Comment by mavasani** [comment]
http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Compiler/MethodCompiler.cs,1078

If the semantic model fetched above was somehow shared between all methods in the same syntax tree, and strongly held in memory until the method bodies in the entire syntax tree have been compiled and analyzed (i.e. used in relevant analyzer callbacks), then semantic model analyzer callbacks would be more optimal.

**Comment by CyrusNajmabadi** [comment]
Note: what i'm asking for is *not* a cache.  I don't want random objects held for random times.  What i'm explicitly asking for is that a single SemanticModel be used during the lifetime of the analysis of a single file.  That file analysis should have a well defined lifetime for diagnostic-analyzers, and thus it should be easy to create the single SemanticModel for that file, and then let go of it when the file analysis is done.

**Comment by CyrusNajmabadi** [comment]
Agree with @mavasani .  Note that the use case the IDE has is for the 'simplify type names' feature.  This is basically the worst case for analyzers.  We effectively have to look at *every* name in a file to see if it's simplifiable.  And this requires semantic queries to determine this.  So we're effectively causing rebinding of practically everything in teh file.  This is obviously not good given that the compiler already just had to go through and do that same binding itself.  We would likely get upward of a 2x speedup here (on our *****slowest***** analyzer by far) by having this.

**Comment by CyrusNajmabadi** [comment]
@mavasani if you try out anything with a PR, i'd be happy to try it out with my current perf work to see if it helps.

**Comment by mavasani** [comment]
Tagging @gafter as I recall discussing similar stuff with him when we added semantic model actions.

@CyrusNajmabadi - I am not sure I am proficient enough on compiler's method binding code to implement this functionality, especially given this can (and probably will) lead to increase in compiler's memory footprint and hence needs quite a bit of performance measurements. I think it would be best for the compiler team to take this up.

**Comment by CyrusNajmabadi** [comment]
>  I think it would be best for the compiler team to take this up.

Makes sense to me.  Note that I would be fine with this being something a driver could opt-into.

**Comment by CyrusNajmabadi** [comment]
@gafter @jaredpar  Do you have any thouhts on how feasible this would be?  To catch you up:

1. IDE scenarios would likely benefit from the compiler using a single semantic model when both processing a file *and* when calling back into analyzers.

Currently, we do not use a single semantic model here.  Compiler creates its own semantic models on-demand for the file, but then creates a fresh one when calling back into registered diagnostic analyers.   This means that all the semantic questions those analyzers call end up duplicating the binding work the compiler just did.

For some of our expensive analyzers we are completely dominated by binding time.  These analyzer would likely see massive gains from reusing a single semantic model here in analyzer scenarios.

This space, unfortunately, is a bit too complex for IDE to prototype out to validate.  We could really use some help with some compiler expertise here to try things out.  If someone from the compiler team could help, we could validate on the IDE side to see what sort of improvement we could get here.

Thanks much!

**Comment by CyrusNajmabadi** [comment]
Tagging @sharwell here as well as he's been doin a lot of investigation and perf fixes in this space.  There's def a ton of low-hanging-fruit that IDE is biting off to make things better.

However, even once we've done that, for some of our analyzers we're seeing so much time spent in binding.  So we're very keep on seeing if there's a driver-level fix that would pay off big.

Note that these scenarios are very impactful as these costs hit users both in terms of large CPU hits as well as excessive garbage production/GCs.

**Comment by mavasani** [comment]
> Currently, we do not use a single semantic model here. Compiler creates its own semantic models on-demand for the file, but then creates a fresh one when calling back into registered diagnostic analyers. This means that all the semantic questions those analyzers call end up duplicating the binding work the compiler just did.

This is not completely correct. Analyzers _do_ share the semantic model with the compiler for all the IOperation callbacks + SyntaxNode callbacks for nodes within executable code. However, for [SemanticModelAction](http://sourceroslyn.io/#q=RegisterSemanticModelAction) callbacks where analyzer intends to analyze semantics of the entire syntax tree, analyzer driver is forced to use a separate semantic model as compiler's method body binding does not re-use the same semantic model for all methods within a syntax tree - it creates one on the fly for each method body (see [here](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Compiler/MethodCompiler.cs,1078)). If the compiler can avoid this and use the same semantic model for binding + analyzing everything in a file, then we can improve the performance of analyzers registering SemanticModelAction by avoiding duplicate binding cost. This is the work being requested by the IDE team to improve the performance of the slowest IDE analyzer (simplify type names).

**Comment by CyrusNajmabadi** [comment]
@mavasani Thanks for making that clear.  That's definitely the case i'm referring to.

**Comment by sharwell** [comment]
@mavasani Today I measured the improvement of using the IOperation semantic model for code blocks in simplify type name. It produces a 10+GB reduction in allocation and a few hundred thread-seconds reduction in CPU usage. We can realize this improvement through the following:

1. Run code block analyzer callbacks _before_ SemanticModel analyzer callbacks
2. Register the individual `MemberSemanticModel` instances with the `SyntaxTreeSemanticModel` used for SemanticModel analyzer callbacks before calling into the analyzer

This will allow the SemanticModel analyzers to share information with other semantic analyzers.

**Comment by sharwell** [comment]
I looked further, and it seems SemanticModel callbacks are already called after code block callbacks. However, caching all the semantic models produces significant memory pressure because the semantic model callbacks are all grouped at the end of analysis. In addition to ensuring the member semantic models are registered with a common syntax tree semantic model, the event processing queue should have the following ordering rules applied to reduce memory pressure:

1. Prefer to process an event for a semantic model that already exists instead of an event for a new syntax tree.
2. Execute semantic model actions after the last code block action for a syntax tree is executed.
